Remove unecessary mask in StrokeHandler

bhennion opened the pull request:
This PR improves performance and mem usage in StrokeHandler for strokes that require a full redraw at each event, by removing an unecessary mask.

It also fixes the display of single-point strokes that require a full redraw, and removes the corresponding warnings.

This is ready for a review.

Technius commented on src/core/control/tools/StrokeHandler.cpp:
```diff
@@ -32,26 +32,24 @@ StrokeHandler::~StrokeHandler() {
-         */
-        cairo_set_operator(crMask, CAIRO_OPERATOR_CLEAR);
-        cairo_paint(crMask);
+    assert(stroke && stroke->getPointCount() > 0);
```
I'm conflicted about this assertion. If I understand correctly, on one hand, we want to avoid having to draw a stroke if the user has not initiated stroke input. On the other hand, it does not seem obvious to me that draw is only invoked in the above situation, which means that a user could get segfaults when drawing stuff.

The assertion can be improved as follows:
```suggestion
    g_assert(stroke && stroke->getPointCount() > 0 && "can only draw a stroke if user has initiated stroke input!");
```

bhennion replied:
I don't think `draw` can ever be called if the stroke has not yet been created. The stroke is created, with an initial point, at the same time as the StrokeHandler is created (see `XojPageView::onButtonPressEvent`).

Technius commented on src/core/control/tools/StrokeHandler.cpp:
```diff
@@ -363,20 +349,39 @@ void StrokeHandler::onButtonDoublePressEvent(const PositionInputData& pos) {
+    const int width = static_cast<int>(std::ceil((visibleRect->width + strokeWidth) * ratio));
+    const int height = static_cast<int>(std::ceil((visibleRect->height + strokeWidth) * ratio));
+
+    surfMask = cairo_image_surface_create(CAIRO_FORMAT_A8, width, height);
```
I think `surfMask` needs to be destroyed first if it already exists, otherwise there may be a memory leak.

Also, would it be better to use `gdk_window_create_similar_image_surface`?

bhennion replied:
I don't think `surfMask` should ever exist before that.
The entire class was structured before the switch to `GTK3`. Before that, the InputHandlers used to survive a single stroke creation: there was one IntpuHandler of each type for each page.
Now, the class is created upon ButtonPressEvent (see `XojPageView::onButtonPressEvent`). An entire refactor of all the InputHandlers would help clearing all that out: someday, maybe.

I'll add an assert.

bhennion replied:
About `gdk_window_create_similar_image_surface`: at this point, we do not have access to any surface to mimic. We only have a surface in the `draw` loop.

Technius commented on src/core/control/tools/StrokeHandler.h:
```diff
@@ -72,6 +72,8 @@ class StrokeHandler: public InputHandler {
     SnapToGridInputHandler snappingHandler;
 
 private:
+    void createMask();
```
This needs a doc comment. Also, I think this should be renamed to some more descriptive like `createStrokeRenderBuffer`. I get that technically we're applying a cairo mask to the page render buffer surface, but that obscures the real intention (we draw the incoming points to a buffer that we can use for full repaint).

Technius commented on src/core/control/tools/StrokeHandler.cpp:
```diff
@@ -363,20 +349,39 @@ void StrokeHandler::onButtonDoublePressEvent(const PositionInputData& pos) {
 
 void StrokeHandler::destroySurface() {
     if (surfMask || crMask) {
+        assert(!this->fullRedraw);
```
What is this assert guarding against?

bhennion replied:
Here for historic reasons :) Will get rid of it

Technius commented on src/core/control/tools/StrokeHandler.cpp:
```diff
@@ -89,8 +87,12 @@ void StrokeHandler::paintTo(const Point& point) {
+                    // Paint the dot for the user to see
+                    this->paintDot(crMask, endPoint.x, endPoint.y, width);
+                }
+                this->redrawable->repaintRect(endPoint.x - 0.5 * width, endPoint.y - 0.5 * width, width, width);
```
Do we need to repaint around the end point if `fullRedraw` is true? The endpoint is only drawn if false.

bhennion replied:
`repaintRect` triggers the call to `draw`, where the `paintDot` will be called in the case we don't use the mask.
I'll add a comment.

Technius commented on src/core/control/tools/StrokeHandler.cpp:
```diff
@@ -330,29 +332,13 @@ void StrokeHandler::onButtonPressEvent(const PositionInputData& pos) {
+        this->createMask();
+        this->paintDot(crMask, this->buttonDownPoint.x, this->buttonDownPoint.y, width);
     }
+    this->redrawable->repaintRect(this->buttonDownPoint.x - 0.5 * width, this->buttonDownPoint.y - 0.5 * width, width,
```
See my above question about full repaint.

Technius commented on src/core/gui/PageView.cpp:
```diff
@@ -853,36 +853,32 @@ void XojPageView::paintPageSync(cairo_t* cr, GdkRectangle* rect) {
     if (this->inputHandler) {
         cairo_save(cr);
-        int dpiScaleFactor = xournal->getDpiScaleFactor();
-        cairo_scale(cr, 1.0 / dpiScaleFactor, 1.0 / dpiScaleFactor);
```
Isn't scaling by `1.0 / dpiScaleFactor` different from scaling by `zoom`?

Technius replied:
Please add a doc comment somewhere that explicitly states this assumption ((maybe in PageView? not sure where an appropriate place would be). It will help greatly when we revisit the code!

Technius reviewed (commented):
I'm not familiar with the stroke rendering and handling code, but I took a look and left some feedback.

* Nice work with handling the zoom more uniformly; this will help avoid some confusion in the future. However, I would like to see doc comments (on the appropriate methods) explaining when we need to multiply by zoom and when we do not need to multiply by zoom.
* Can you clarify what you mean by "the corresponding warnings" that occur for single point strokes?

Further comments on `StrokeHandler.{cpp,h}` in general (no need to take action on these):
* I don't really understand what the `fullRedraw` flag is and when it is set. We should add doc comments that explain the logic in the future.
* We need to explain what the "masking surface" is in more detail. I had to read through the file multiple times to understand that it is a temporary buffer used to render the stroke that is currently being drawn.

bhennion commented:
Thanks for the feedback!

> I'm not familiar with the stroke rendering and handling code, but I took a look and left some feedback.
> 
> * Nice work with handling the zoom more uniformly; this will help avoid some confusion in the future. However, I would like to see doc comments (on the appropriate methods) explaining when we need to multiply by zoom and when we do not need to multiply by zoom.

I'll try and add comments. Basically, the idea is to keep the following assumption true:

- The cairo context fed to `draw`/`paint` is always in page coordinates (i.e. not further scaling or offset is ever needed to paint something in `draw`).

This all thing could be simplified by applying a `cairo_surface_set_device_scale` on the `XojPageView::crBuffer` upon creation, and removing most of the `cairo_scale`'s. It's however not quite that simple (because the zoom changes), and the necessary changes for that go to deep into the view refactor. I'd rather leave that either to a later PR, or to @LittleHuba.

> * Can you clarify what you mean by "the corresponding warnings" that occur for single point strokes?

Try starting a filled stroke, and look at the terminal output. You should get some 
```
** (xournalpp:29090): WARNING **: 08:54:56.617: DocumentView::drawStroke empty stroke...
```
This is due to a view.drawStroke being called on a stroke with less than 2 points.

> Further comments on `StrokeHandler.{cpp,h}` in general (no need to take action on these):
> 
> * I don't really understand what the `fullRedraw` flag is and when it is set. We should add doc comments that explain the logic in the future.
> * We need to explain what the "masking surface" is in more detail. I had to read through the file multiple times to understand that it is a temporary buffer used to render the stroke that is currently being drawn.

bhennion commented:
I added a new subclass StrokeHandler::Mask to handler construction and destruction of the mask safely. The mask is now an std::optional<Mask>. The existence of a mask also replaces the flag fullRedraw (== !mask).

Also took care of your other comments (I think).

bhennion commented:
I rebased on the latest master.

I also simplified further the zoom handling in `PageView::paintPageSync`. Now, the VerticalSpaceTool is fed, like every other tool, a cairo context set up in page coordinates.

@Technius You can go a rereview it all, whenever you have time.

LittleHuba commented:
> This all thing could be simplified by applying a cairo_surface_set_device_scale on the XojPageView::crBuffer upon creation, and removing most of the cairo_scale's. It's however not quite that simple (because the zoom changes), and the necessary changes for that go to deep into the view refactor. I'd rather leave that either to a later PR, or to @LittleHuba.

I think you are doing an awesome job here. It is also a very good idea to keep your PRs small. Better have five small incremental PRs than one big PR. They will be easier to review and faster to merge. Then I see no reason why you couldn't approach this task yourself :slightly_smiling_face:

bhennion commented:
> I think you are doing an awesome job here. It is also a very good idea to keep your PRs small. Better have five small incremental PRs than one big PR. They will be easier to review and faster to merge. Then I see no reason why you couldn't approach this task yourself slightly_smiling_face

I'll try to keep the PRs contained :-)

Technius reviewed (approved):
LGTM except one minor comment. I did not have time to test, but I didn't see any obvious problems with the code.

bhennion commented:
Thanks @Technius for the review!

I added a comment in PageView. Ideally (maybe in a follow up PR, someday), I think all those painters should inherit from a common class, where this comment should be moved.

Merging in 24 hours unless an objection is raised.
